mlir-sycl/lib/Dialect/SYCL/Transforms/ConstantPropagation.cpp

```diff
@@ -91,12 +112,630 @@ class ConstantArithArg : public ConstantExplicitArg {
+  Value zero = builder.create<arith::ConstantIndexOp>(loc, 0);
+  Value allocatedValue =
+      getNewValue(builder, loc, info, MemRefType::get(1, type));
+  Value value = builder.create<memref::LoadOp>(loc, type, allocatedValue, zero);
+  oldValue.replaceAllUsesWith(value);
+  LLVM_DEBUG({
+    llvm::dbgs().indent(8) << "Replaced:\n";
+    llvm::dbgs().indent(10) << oldValue << "\n";
+    llvm::dbgs().indent(8) << "with:\n";
+    llvm::dbgs().indent(10) << zero << "\n";
+    for (Value operand : allocatedValue.getDefiningOp()->getOperands())
+      llvm::dbgs().indent(10) << operand << "\n";
+    llvm::dbgs().indent(10) << allocatedValue << "\n";
+    llvm::dbgs().indent(10) << value << "\n";
+  });
+  op->erase();
+}
+
+template <>
+Value ConstantSYCLGridArgs::NumWorkItemsRewriter::getNewValue(
+    OpBuilder &builder, Location loc, const NDRInfo &info, Type type) const {
+  return info.getConstantGlobalSizeValue(builder, loc, type);
+}
+
+template <>
+Value ConstantSYCLGridArgs::NumWorkGroupsRewriter::getNewValue(
+    OpBuilder &builder, Location loc, const NDRInfo &info, Type type) const {
+  const polygeist::NDRangeInformation &ndrInfo = info.getNDRange();
+  return createIDRange<SYCLRangeConstructorOp>(
+      builder, loc, type,
+      llvm::map_range(llvm::zip(ndrInfo.getGlobalSizeInfo().getConstantValues(),
+                                ndrInfo.getLocalSizeInfo().getConstantValues()),
+                      [&](auto iter) {
+                        auto [globalSize, localSize] = iter;
+                        return globalSize / localSize;
+                      }));
+}
+
+template <>
+Value ConstantSYCLGridArgs::WorkGroupSizeRewriter::getNewValue(
+    OpBuilder &builder, Location loc, const NDRInfo &info, Type type) const {
+  return info.getConstantLocalSizeValue(builder, loc, type);
+}
+
+template <>
+Value ConstantSYCLGridArgs::GlobalOffsetRewriter::getNewValue(
+    OpBuilder &builder, Location loc, const NDRInfo &info, Type type) const {
+  return info.getConstantOffsetValue(builder, loc, type);
+}
+
+//===----------------------------------------------------------------------===//
+// ConstantImplicitArgBase
+//===----------------------------------------------------------------------===//
+
+FunctionOpInterface
+ConstantImplicitArgBase::cloneFunction(FunctionOpInterface original,
+                                       CallOpInterface call, Twine callPath) {
+  OpBuilder functionCloner(original);
+  auto clone = cast<FunctionOpInterface>(functionCloner.clone(*original));
+  clone->setAttr(
+      SymbolTable::getSymbolAttrName(),
+      functionCloner.getAttr<StringAttr>(
+          Twine("__").concat(original.getName()).concat(callPath).str()));
+  call.setCalleeFromCallable(
+      FlatSymbolRefAttr::get(static_cast<SymbolOpInterface>(clone)));
+  return clone;
+}
+
+LogicalResult ConstantImplicitArgBase::match(Operation *op) const {
+  return TypeSwitch<const ConstantImplicitArgBase *, LogicalResult>(this)
```

Comment:
Nit: I assume the `TypeSwitch` here is a preparation for further extensions; otherwise I'd prefer a simpler `cast<ConstantSYCLGridArgs>` here.

Reply:
Yep, in fact, I'm planning to merge explicit-implicit code so this would help.